core/modules/loader/CentralWorker.cc

```diff
@@ -0,0 +1,1114 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2018 AURA/LSST.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+
+// Class header
+#include "loader/CentralWorker.h"
+
+// system headers
+#include <boost/asio.hpp>
+#include <iostream>
+
+// qserv headers
+#include "loader/BufferUdp.h"
+#include "loader/CentralWorkerDoListItem.h"
+#include "loader/LoaderMsg.h"
+#include "proto/loader.pb.h"
+#include "proto/ProtoImporter.h"
+
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.loader.CentralWorker");
+}
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+
+CentralWorker::CentralWorker(boost::asio::io_service& ioService_,
+    std::string const& masterHostName_,   int masterPort_,
+    int threadpoolSize_, int sleepTime_,
+    std::string const& hostName_,         int udpPort_,
+    boost::asio::io_context& io_context_, int tcpPort_)
+    : Central(ioService_, masterHostName_, masterPort_, threadpoolSize_, sleepTime_),
+      _hostName(hostName_), _udpPort(udpPort_),
+      _tcpPort(tcpPort_), _ioContext(io_context_) {
+    _server = std::make_shared<WorkerServer>(ioService, _hostName, _udpPort, this);
+    _tcpServer = std::make_shared<ServerTcpBase>(_ioContext, _tcpPort, this);
+    _tcpServer->runThread();
+    _startMonitoring(); // This must be the last item in the constructor.
+}
+
+CentralWorker::~CentralWorker() {
+    _wWorkerList.reset();
+    _tcpServer.reset();
```

Comment:
I think you choose to reset these `Ptr`s because they have pointers to `this`?
it looks like `_tcpServer` is never passed anywhere else in which case it will be destroyed when calling `reset`; but maybe you should put it in a `unique_ptr` to guarantee?
However `_wWorkerList` is passed into the `DoList` and put on a queue, which could lead to trouble. Maybe you should unregister `this` from that object before destroying `this` instance?

Reply:
what about `_centralWorkerDoListItem`?

Reply:
Central items need to live until the program terminates. Getting these programs to terminate cleanly can be done but is not trivial. I think it is better to spend effort getting the primary functionality working than getting a clean shutdown.